selfservice/strategy/code/strategy_login.go

```diff
@@ -129,157 +124,142 @@ func (s *Strategy) Login(w http.ResponseWriter, r *http.Request, f *login.Flow,
 		decoderx.MustHTTPRawJSONSchemaCompiler(loginMethodSchema),
 		decoderx.HTTPDecoderAllowedMethods("POST"),
 		decoderx.HTTPDecoderJSONFollowsFormFormat()); err != nil {
-		return nil, s.HandleLoginError(w, r, f, &p, err)
+		return nil, s.HandleLoginError(r, f, &p, err)
 	}
 
-	if err := flow.EnsureCSRF(s.deps, r, f.Type, s.deps.Config().DisableAPIFlowEnforcement(r.Context()), s.deps.GenerateCSRFToken, p.CSRFToken); err != nil {
-		return nil, s.HandleLoginError(w, r, f, &p, err)
+	if err := flow.EnsureCSRF(s.deps, r, f.Type, s.deps.Config().DisableAPIFlowEnforcement(ctx), s.deps.GenerateCSRFToken, p.CSRFToken); err != nil {
+		return nil, s.HandleLoginError(r, f, &p, err)
 	}
 
-	codeManager := NewCodeStateManager(f, s, &p)
+	// By Default the flow should be in the 'choose method' state.
+	SetDefaultFlowState(f, p.Resend)
 
-	codeManager.SetCreateCodeHandler(func(ctx context.Context, f *login.Flow, strategy *Strategy, p *updateLoginFlowWithCodeMethod) error {
-		strategy.deps.Audit().
-			WithSensitiveField("identifier", p.Identifier).
-			Info("Creating login code state.")
-
-		// Step 1: Get the identity
-		i, cred, err := strategy.getIdentity(ctx, p.Identifier)
+	switch f.GetState() {
+	case flow.StateChooseMethod:
+		if err := s.loginSendEmail(ctx, w, r, f, &p); err != nil {
+			return nil, s.HandleLoginError(r, f, &p, err)
+		}
+		return nil, nil
+	case flow.StateEmailSent:
+		i, err := s.loginVerifyCode(ctx, r, f, &p)
 		if err != nil {
-			return err
+			return nil, s.HandleLoginError(r, f, &p, err)
 		}
+		return i, nil
+	case flow.StatePassedChallenge:
+		return nil, s.HandleLoginError(r, f, &p, errors.WithStack(schema.NewNoLoginStrategyResponsible()))
+	}
```

Comment:
IMO, this is easier to read and understand what's going on. Is state choose_method? Do this. Is state passed_challenge? Do that.

With the callbacks, it wasn't very clear what is executed when, because `CreateCodeHandler` does not indicate that the state is `choose_method`. Plus, we save a lot of code complexity :)